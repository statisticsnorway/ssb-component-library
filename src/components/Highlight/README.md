# Highlight

> A component for highlighting words, parts of a text line or paragraph.

### Usage

#### HTML

```html
<mark class="ssb-mark">Default text here</mark>

<mark class="ssb-mark negative">Negative text here</mark>
```

#### React

```jsx harmony
<Highlight>Default text here</Highlight>

<Highlight negative>Negative text here</Highlight>


```

Available props:

| Name      | Type             | Description                |
| --------- | ---------------- | -------------------------- |
| children  | Required node    | Content within the wrapper |
| className | string           | Optional container class   |
| negative  | Optional boolean | Styling for dark theme     |
