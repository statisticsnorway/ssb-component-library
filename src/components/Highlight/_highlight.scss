.ssb-mark {
  @include open-sans;
  background-color: $ssb-green-1;
  font-style: normal;
  letter-spacing: normal;
  line-height: 1.6;

  &.negative {
    background-color: $ssb-green-5;
    color: $ssb-white;
  }
}

// code below is added for screen reader support and makes the screen readers announce where highlight starts and ends.
// Todo: remove when support for <mark> is more widespread among screenreaders

mark::before,
mark::after {
  clip-path: inset(100%);
  clip: rect(1px, 1px, 1px, 1px);
  height: 1px;
  overflow: hidden;
  position: absolute;
  white-space: nowrap;
  width: 1px;
}

mark::before {
  content: ' [highlight start] ';
}

mark::after {
  content: ' [highlight end] ';
}
