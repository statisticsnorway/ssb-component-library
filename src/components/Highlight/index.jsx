import React from 'react';
import PropTypes from 'prop-types';

const Highlight = ({
	children, className, negative,
}) => (
	<mark className={`ssb-mark${negative ? ' negative' : ''}${className ? ` ${className}` : ''}`}> { children } </mark>
);

Highlight.defaultProps = {
	negative: false,
};

Highlight.propTypes = {
	children: PropTypes.node.isRequired,
	className: PropTypes.string,
	negative: PropTypes.bool,
};

export default Highlight;
