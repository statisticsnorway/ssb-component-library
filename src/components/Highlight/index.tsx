import React, { type ReactNode } from 'react'
import classNames from '../../utils/utils'

interface HighlightProps {
  children: ReactNode
  className?: string
  negative?: boolean
}

const Highlight: React.FC<HighlightProps> = ({ children, className, negative = false }) => (
  <mark className={classNames('ssb-mark', negative && 'negative', className)}> {children} </mark>
)

export default Highlight
