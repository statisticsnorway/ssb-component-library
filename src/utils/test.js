import fs from 'fs';
import path from 'path';
import { render as rtlRender } from '@testing-library/react';

export * from '@testing-library/react';
export * from '@testing-library/jest-dom';

const styles = fs.readFileSync(path.relative(process.cwd(), './lib/bundle.css'), 'utf8');

export function render(ui, options) {
	const view = rtlRender(ui, options);

	const styleElement = document.createElement('style');
	styleElement.innerHTML = styles;
	document.body.appendChild(styleElement);

	return view;
}
