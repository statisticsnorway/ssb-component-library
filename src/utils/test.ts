import fs from 'fs'
import path from 'path'
import { RenderOptions, RenderResult, render as rtlRender } from '@testing-library/react'
import { ReactNode } from 'react'

export * from '@testing-library/react'

const styles = fs.readFileSync(path.relative(process.cwd(), './lib/bundle.css'), 'utf8')

export function render(ui: ReactNode, options?: RenderOptions): RenderResult {
  const view = rtlRender(ui, options)

  const styleElement = document.createElement('style')
  styleElement.innerHTML = styles
  document.body.appendChild(styleElement)

  return view
}
