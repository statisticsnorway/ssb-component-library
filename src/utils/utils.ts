// utils.ts
const classNames = (...classes: (string | boolean | undefined)[]): string => classes.filter(Boolean).join(' ')

export default classNames
