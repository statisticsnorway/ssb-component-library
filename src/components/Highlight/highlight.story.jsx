import React from 'react'
import Highlight from './index'

const negativeStyle = {
  backgroundColor: '#274247',
  padding: '100px',
}

export default {
  title: 'Highlight',
}

export const Default = () => (
  <div style={{ padding: '100px' }}>
    <p>
      «Ser du etter en hytte med slik beliggenhet, må du regne med å måtte ut med en bra slump med penger
      <Highlight>hvis du ikke er så heldig å arve</Highlight>– selv om også dét kan være dyrt.»
    </p>
  </div>
)

export const Negative = () => (
  <div style={negativeStyle}>
    <p>
      «Ser du etter en hytte med slik beliggenhet, må du regne med å måtte ut med en bra slump med penger
      <Highlight negative>hvis du ikke er så heldig å arve</Highlight>– selv om også dét kan være dyrt.»
    </p>
  </div>
)
