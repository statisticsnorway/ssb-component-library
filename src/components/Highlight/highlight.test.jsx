import React from 'react'
import { render } from '../../utils/test'
import Highlight from './index'

describe('Highlight component', () => {
  test('Matches the snapshot', () => {
    const { asFragment } = render(<Highlight>«Highlight»</Highlight>)
    expect(asFragment()).toMatchSnapshot()
  })
  test('Toggles classNames correctly', () => {
    const { asFragment } = render(<Highlight negative>«Highlight»</Highlight>)
    expect(asFragment()).toMatchSnapshot()
  })
})
