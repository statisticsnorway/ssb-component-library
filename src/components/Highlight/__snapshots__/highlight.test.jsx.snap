// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Highlight component Matches the snapshot 1`] = `
<DocumentFragment>
  <mark
    class="ssb-mark"
  >
     «Highlight» 
  </mark>
</DocumentFragment>
`;

exports[`Highlight component Toggles classNames correctly 1`] = `
<DocumentFragment>
  <mark
    class="ssb-mark negative"
  >
     «Highlight» 
  </mark>
</DocumentFragment>
`;
